calyx-opt/src/passes/top_down_compile_control.rs

```diff
@@ -388,195 +475,178 @@ impl<'b, 'a> Schedule<'b, 'a> {
+                    let state_guard = Self::build_query(
+                        self.builder,
+                        used_slicers,
+                        &fsm_rep,
+                        fsm,
+                        &signal_on,
+                        &state,
+                        &fsm_size,
+                    );
+                    assigns.iter_mut().for_each(|asgn| {
+                        asgn.guard.update(|g| g.and(state_guard.clone()))
+                    });
+                    assigns
                 }),
         );
 
         // transition assignments
+        // the following updates are meant to ensure agreement between the two
+        // fsm registers; hence, all registers must be updated if `duplicate` is chosen
         group.borrow_mut().assignments.extend(
             self.transitions.into_iter().flat_map(|(s, e, guard)| {
-                let (end_const, trans_guard) = match encoding {
-                    Encoding::Binary => {
-                        structure!(self.builder;
-                            let end_const = constant(e, fsm_size);
-                            let start_const = constant(s, fsm_size);
-                        );
-                        let trans_guard =
-                            guard!((fsm["out"] == start_const["out"]) & guard);
-
-                        (end_const, trans_guard)
-                    }
-                    Encoding::OneHot => {
-                        let end_constant_value = u64::pow(
-                            2,
-                            e.try_into().expect("failed to convert to u32"),
-                        );
-
-                        let trans_guard = Self::build_one_hot_query(
-                            self.builder,
-                            &mut used_slicers,
-                            &fsm,
-                            &signal_on,
-                            &s,
-                            &fsm_size,
-                        );
-                        let end_const = self
-                            .builder
-                            .add_constant(end_constant_value, fsm_size);
-
-                        (end_const, trans_guard.and(guard))
-                    }
-                };
-                let ec_borrow = end_const.borrow();
-                vec![
-                    self.builder.build_assignment(
-                        fsm.borrow().get("in"),
-                        ec_borrow.get("out"),
-                        trans_guard.clone(),
-                    ),
-                    self.builder.build_assignment(
-                        fsm.borrow().get("write_en"),
-                        signal_on.borrow().get("out"),
-                        trans_guard,
+                // get a transition guard for the first fsm register, and apply it to every fsm register
+                let state_guard = Self::build_query(
```

Comment:
Will picking the first register lead to uneven query-ing?

Reply:
As of now, there's only one duplicate register, so I think using the output of the first register's logic for the second register shouldn't be an issue. If we end up needing more duplicate registers, then that might create some bottleneck, but I think we should be good for now?